benchmarks/drop/drop.py

```diff
@@ -0,0 +1,381 @@
+inspect eval drop/drop.py
+
+# run with more or no fewshot samples
+inspect eval drop/drop.py -T fewshot=5
+inspect eval drop/drop.py -T fewshot=false
+"""
+
+import re
+import string
+from typing import Dict, List, Tuple, Union
+
+import numpy as np
+from scipy.optimize import linear_sum_assignment
+
+from inspect_ai import Task, task
+from inspect_ai.dataset import Sample, hf_dataset
+from inspect_ai.scorer import Score, Target, bootstrap_std, mean, scorer
+from inspect_ai.solver import (
+    Generate,
+    TaskState,
+    generate,
+    prompt_template,
+    solver,
+    system_message,
+)
+
+DROP_PROMPT_TEMPLATE = {
+    # Based on the prompt provided here: https://github.com/openai/simple-evals/blob/main/drop_eval.py#L261C13-L283C91
+    "system": """You will be asked to read a passage and answer a question.\n\n""",
+    "system_w_examples": """You will be asked to read a passage and answer a question. Some examples of passages and Q&A are provided below.\n\n"""
+    + """Examples\n{examples}\n\n""",
+    "user": """Your Task\n---\n{prompt}\n\n"""
+    + """Think step by step, then write a line of the form "Answer: $ANSWER" at the end of your response.\n""",
+}
+
+# Borrowed from: https://github.com/openai/simple-evals/blob/main/common.py#L24
+ANSWER_PATTERN = r"(?i)Answer\s*:\s*([^\n]+)"
+
+
+@task
+def drop(
+    fewshot: int = 3,
+    fewshot_seed: int = 42,
+) -> Task:
+    """Inspect task implementing the DROP benchmark.
+
+    Arguments:
+        fewshot (int): Number of few shot examples to use.
+        fewshot_seed (int): Random seed for sampling few shot examples.
+    """
+    # Get prompt
+    prompt_dict = DROP_PROMPT_TEMPLATE
+    prompt_user = prompt_dict["user"]
+    prompt_sys = prompt_dict["system"]
+    prompt_sys_w_examples = prompt_dict["system_w_examples"]
+
+    # Add system message based on zero-shot or few-shot prompting
+    if fewshot:
+        fewshot_samples = hf_dataset(
+            "EleutherAI/drop",
+            split="train",
+            trust=True,
+            sample_fields=record_to_sample,
+            shuffle=True,
+            seed=fewshot_seed,
+            limit=fewshot,
+        )
+        sys_msg = system_message(
+            prompt_sys_w_examples.format(
+                examples="\n\n".join(
+                    [sample_to_fewshot(sample) for sample in fewshot_samples]
+                )
+            )
+        )
+    else:
+        sys_msg = system_message(prompt_sys)
+
+    plan = [
+        sys_msg,
+        prompt_template(prompt_user),
+        generate(),
+        drop_output_parser(),
+    ]
+
+    return Task(
+        dataset=hf_dataset(
+            "EleutherAI/drop",
+            split="validation",
+            trust=True,
+            sample_fields=record_to_sample,
+        ),
+        plan=plan,
+        scorer=drop_f1_scorer(),
+    )
+
+
+@solver
+def drop_output_parser():
+    async def solve(state: TaskState, generate: Generate) -> TaskState:
```

Comment:
I think we want all of this to happen in the scorer (as it has full access to the task state as well). That way the final completion text is still clearly visible in the message history. Could be missing some important reason why things are this way though, let's discuss.

Reply:
A separate solver for parsing output allowed us to isolate some model specific logic for the parser, but it was removed later. So it makes sense to integrate this logic in the scorer itself. Refactored code accordingly.